### PR TITLE
feat: support HashiCorp Vault with token based access

### LINKED ISSUE #20: Support HashiCorp Vault as source system
- [ ] Add basic support for HashiCorp Vault with token based auth #38
- [ ] Add support for kubernetes auth

### PR DESCRIPTION
**Pull request by florianrusch**
fixes #20

### CODE REVIEW

File: cmd/backup.go [hunk 1]
```diff
@@ -79,3 +81,12 @@ var backupCmd = &cobra.Command{
 		return nil
 	},
 }
+
+func init() {
+	rootCmd.PersistentFlags().StringVar(
```
Comment (a-mesin) [inline cmd/backup.go:86]: Shouldnt this flag be used only with `backupCmd` or do we really need the config flag on every subcommand?

File: config.yaml [hunk 2]
```diff
@@ -1,9 +1,15 @@
+# In order to enable environmental substitution, it is necessary to include all potential configurations in this
+# configuration file. Otherwise, the environment variables will not be automatically added to the ResourceConfig
+# (`mapstructure:",remain"`).
+
+loglevel: info
+
 source:
-  type: filesystem
-  path: /tmp/source.txt
+  type:
```
Comment (a-mesin) [inline config.yaml:8]: I somehow don't like mixing up the config for the different types on the same level as the user might get confused on what is required for the particular type. We should at least make sure to document which properties are required for the different types somewhere. 

Another alternative would be to introduce an additional level, like:
```yaml
source:
  type: filesystem
  filesystem:
    path: /tmp/foo.txt
  vault:
    token:
    address:
```

What do you think?

File: internal/source/hashicorpvault/config.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+const Type = "hashicorpvault"
+
+type config struct {
+	Address string `validate:"http_url,required"`
+	Token   string `validate:"required"`
+}
+
+var validate *validator.Validate
+
+// newConfig provides the specific `config` struct. Therefor it takes the generic `globalConfig.ResourceConfig` and
```
Comment (a-mesin) [inline internal/source/hashicorpvault/config.go:36]:
```suggestion
// newConfig provides the specific `config` struct. It takes the generic `globalConfig.ResourceConfig` and
```

File: internal/source/hashicorpvault/config.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	return &c, nil
+}
+
+// LogValue customizes how the `config` struct will be printed in the logs.
+func (c *config) LogValue() slog.Value {
+	return slog.GroupValue(slog.String("address", c.Address), slog.String("token", "***"))
+}
+
+// GetHashiCorpVaultConfig was implement in regard to the
+// `vault.DefaultConfig()` method. While the implementation the
```
Comment (a-mesin) [inline internal/source/hashicorpvault/config.go:61]: I am not sure if I get the second sentence there

File: internal/source/hashicorpvault/config.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	return &c, nil
+}
+
+// LogValue customizes how the `config` struct will be printed in the logs.
+func (c *config) LogValue() slog.Value {
+	return slog.GroupValue(slog.String("address", c.Address), slog.String("token", "***"))
+}
+
+// GetHashiCorpVaultConfig was implement in regard to the
```
Comment (a-mesin) [inline internal/source/hashicorpvault/config.go:60]:
```suggestion
// GetHashiCorpVaultConfig is implemented in regard to the
```

File: internal/source/hashicorpvault/reader.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	globalConfig "github.com/cluetec/lifeboat/internal/config"
+	vault "github.com/hashicorp/vault/api"
+	"io"
+	"log/slog"
+)
+
+const snapshotPath = "/sys/storage/raft/snapshot"
+
+// Reader implements the `io.ReaderClose` interface for read the backup from HashiCorp Vault.
```
Comment (a-mesin) [inline internal/source/hashicorpvault/reader.go:28]:
```suggestion
// Reader implements the `io.ReaderClose` interface in order to read the backup from HashiCorp Vault.
```

File: internal/source/hashicorpvault/reader.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+		}
+
+		r.reader = resp.Body
+	}
+
+	return r.reader.Read(b)
+}
+
+func (r *Reader) Close() error {
+	slog.Debug("closing HashiCorp Vault reader")
```
Comment (a-mesin) [inline internal/source/hashicorpvault/reader.go:70]: maybe we should stay consistent in how we write "Hashicorp Vault". In the Read method it is written in lower case.

File: samples/HashiCorp-Vault/README.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
+# Sample: Backup HashiCorp Vault
+
+In this example we will show you how to back up an HashiCorp Vault instance.
```
Comment (a-mesin) [inline samples/HashiCorp-Vault/README.md:3]:
```suggestion
In this example we will show you how to backup an HashiCorp Vault instance.
```

File: samples/HashiCorp-Vault/README.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+## Requirements
+
+The Vault instance needs to use [raft](https://developer.hashicorp.com/vault/docs/configuration/storage/raft) as the
+underlying storage engine.
+
+## Policy
+
+In this sample we are using for simplicity reasons the "root token" to authorize us. In a real world scenario you
+would use a separate took or any other
```
Comment (a-mesin) [inline samples/HashiCorp-Vault/README.md:13]:
```suggestion
would use any other
```

File: samples/HashiCorp-Vault/init-and-fill-vault-with-data.sh [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+do
+    re="Initial Root Token: "
+    if [[ "${SINGLELINE}" =~ $re ]]; then
+        rootToken=$(echo "${SINGLELINE}" | sed "s/${re}//")
+    fi
+done << EOF
+$initOutput
+EOF
+
+# Write unseal keys und token into separate files
```
Comment (a-mesin) [inline samples/HashiCorp-Vault/init-and-fill-vault-with-data.sh:82]:
```suggestion
# Write unseal keys and token into separate files
```

File: samples/HashiCorp-Vault/README.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+The HashiCorp Vault setup is based on a docker compose setup.
+
+```shell
+# -d is starting the container in the background
+$ docker-compose up -d
+```
+
+As we can't simply use the dev mode of Vault, we need to initialize and unseal it first. For this purpose the
+`docker-compose.yaml` contains next to the vault container an additional one which is called `vault-init`. This
```
Comment (a-mesin) [inline samples/HashiCorp-Vault/README.md:38]:
```suggestion
`docker-compose.yaml` contains an additional container which is called `vault-init`. This
```

**Review by a-mesin — CHANGES_REQUESTED** [review]

File: config.yaml (hunk 2 above)
Comment (florianrusch) [reply config.yaml:8]: Thats a good idea. Let's do that ;-)

File: cmd/backup.go (hunk 1 above)
Comment (florianrusch) [reply cmd/backup.go:86]: Good remark, I will move it to the `backupCmd`!

File: internal/source/hashicorpvault/reader.go (hunk 7 above)
Comment (florianrusch) [reply internal/source/hashicorpvault/reader.go]: I fixed that in a better way (I guess 😄)

Have a look at this commit: [`b16a37b` (#38)](https://github.com/cluetec/lifeboat/pull/38/commits/b16a37b7ea7ff81334b1a4c9fe56bef0968a85c8)

**Review by a-mesin — APPROVED** [review]

**Pull request merged**